97b gtr baseline

Juan-Mateos opened the pull request:
Closes #97 

Read instructions [here](https://github.com/nestauk/ai_genomics/tree/97b_gtr_baseline/ai_genomics/pipeline).

The PR includes a new getter to read a gtr projects table including baseline (AI / genomics) projects locally or from s3.

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

Commit:
script to produce and save table with AI genomics GtR projects

Commit:
fixed bugs in script and updated readme

Commit:
added gtr project and baseline project getter

Commit:
changed gtr utils import

Commit:
gtr utils script

india-kerle commented on ai_genomics/utils/reading.py:
```diff
@@ -9,3 +11,13 @@ def read_json(data: Union[pathlib.Path, str]) -> List[Dict]:
         return json.load(json_file)
+
+
+def fetch_s3(s3_path) -> Union[pd.DataFrame, Dict]:
```
we have a function that does this also in ai_genomics.getters.data_getters (load_s3_data) (and indeed similar script of loading and saving data)- you may just want to use that one instead

india-kerle commented on ai_genomics/pipeline/gtr/gtr_utils.py:
```diff
@@ -0,0 +1,22 @@
+    s3 = boto3.resource("s3")
+    ai_genomics_bucket = s3.Bucket("ai-genomics")
+
+    logging.info(f"Fetching table {table_name}")
```
does it matter that we are using logging instead of from `ai_genomics import logger`? I'm not super familiar with all the functionality of logging so just curious!!

india-kerle commented on ai_genomics/pipeline/gtr/gtr_utils.py:
```diff
@@ -0,0 +1,22 @@
+
+    logging.info(f"Fetching table {table_name}")
+
+    return pipe(
```
you could use the load_s3_data function in ai_genomics.getters.data_getters to load gtr

india-kerle replied:
but still have this as a wrapper function i.e.

```
from ai_genomics import bucket_name

def fetch_gtr(table_name: str) -> List[Dict]:
 """Fetch a json gtr object"""
return load_s3_data(bucket_name, f"inputs/gtr/{table_name}.json)
```

india-kerle commented on ai_genomics/pipeline/gtr/make_gtr_projects.py:
```diff
@@ -0,0 +1,177 @@
+
+
+def send_output_to_s3(file_path: str, s3_destination: str):
+    """TODO: PUT THIS IN UTILS"""
```
I reckon you could use save_to_s3 from ai_genomics.getters.data_getters :)

india-kerle commented on ai_genomics/pipeline/gtr/make_gtr_projects.py:
```diff
@@ -0,0 +1,177 @@
+        project_examples["abstractText"].str[:700] + "..."
+    )
+    project_examples = project_examples[["Project title", "Abstract (truncated)"]]
+    project_examples.to_markdown(f"{PROJECT_DIR}/outputs/gtr_examples.md", index=False)
```
wow didn't know about this .to_markdown - very cool !

india-kerle replied:
ah you need to `pip install tabulate` to make this work

india-kerle reviewed (commented):
great! thanks so much for this juan - the comments are largely around using existing functions in ai_genomics.getters.data_getters to load and save data to s3.

india-kerle commented on ai_genomics/pipeline/gtr/make_gtr_projects.py:
```diff
@@ -0,0 +1,177 @@
+
+from ai_genomics.pipeline.gtr.gtr_utils import fetch_gtr
+from ai_genomics import config, PROJECT_DIR
+
```
you forgot some imports -- 

```suggestion
import json
from typing import List, Dict
```

india-kerle commented on ai_genomics/pipeline/gtr/make_gtr_projects.py:
```diff
@@ -0,0 +1,177 @@
+    )
+
+    send_output_to_s3(
+        str(GTR_OUTPUTS_DIR / GTR_PROJ_NAME), f"crunchbase/{GTR_PROJ_NAME}"
```
you're saving this to the crunchbase directory, not GTR:

```suggestion
        str(GTR_OUTPUTS_DIR / GTR_PROJ_NAME), f"gtr/{GTR_PROJ_NAME}"
```

india-kerle commented on ai_genomics/pipeline/gtr/make_gtr_projects.py:
```diff
@@ -0,0 +1,177 @@
+if __name__ == "__main__":
+    logging.info("Reading GtR data")
+    projects = fetch_gtr("gtr_projects-projects")
+    topics = fetch_gtr("gtr_projects-topic")
```
I had issues running this script so needed to:

```suggestion
    topics = pd.DataFrame(fetch_gtr("gtr_projects-topic"))
```

india-kerle commented on ai_genomics/pipeline/gtr/make_gtr_projects.py:
```diff
@@ -0,0 +1,177 @@
+    topics = fetch_gtr("gtr_projects-topic")
+
+    # Topic distribution
+    topic_distr = pd.Series(Counter([topic["text"] for topic in topics]))
```
```suggestion
topic_distr = pd.Series(Counter(topics.text))
```

india-kerle commented on ai_genomics/pipeline/gtr/make_gtr_projects.py:
```diff
@@ -0,0 +1,177 @@
+        config[f"gtr_{disc}_concepts"] for disc in ["ai", "genom"]
+    ]
+
+    ai_projs, genom_projs = [
+        set(
+            [
+                element["project_id"]
+                for element in topics
+                if any(t in element["text"] for t in terms)
+            ]
+        )
+        for terms in [ai_topics, genomics_topics]
+    ]
```
```suggestion
ai_projs, genom_projs = topics[topics.text.isin(ai_topics)].project_id, topics[topics.text.isin(genomics_topics)].project_id
```

Commit:
updated scripts

Commit:
changed getter

Juan-Mateos commented:
I have addressed all comments above. 

`tabulate` is already in `requirements.txt` - sorry I forgot to mention the need to update those.

Thanks for picking up on the `topics` issue - the structure of the json changed between implementations! I fixed this within a list comprehension paradigm :-D

I have rerun the code and everything runs ok.

india-kerle commented on ai_genomics/getters/gtr.py:
```diff
@@ -57,4 +56,4 @@ def get_ai_genomics_project_table(local: bool = True) -> pd.DataFrame:
 
     else:
-        return fetch_s3(f"outputs/gtr/{GTR_PROJ_NAME}")
+        return load_s3_data("ai-genomics", f"outputs/gtr/{GTR_PROJ_NAME}")
```
you can just use BUCKET_NAME as you've imported it instead of "ai-genomics" but its much of a muchness as this works too

india-kerle commented:
thanks for making those changes! Looks good - you can just use from ai_genomics import bucket_name and use bucket_name instead of "ai-genomics" but its much of a muchness as 'ai-genomics' works too.
